[AOTI] Remove aoti_torch_cpu__weight_int4pack_mm_cpu_tensor

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148907

Summary: shim.h is only meant for generic tensor util shim functions. We should switch to use the auto fallback generation, but it will need some extra care on the op schema.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

desertfire committed:
[AOTI] Move aoti_torch_cpu__weight_int4pack_mm_cpu_tensor from shim.h to shim_mkldnn.h

Summary: shim.h is only meant for generic tensor util shim functions. Also aoti_torch_cpu__weight_int4pack_mm_cpu_tensor wasn't declared as a C API.

[ghstack-poisoned]

desertfire commented:
xref https://github.com/pytorch/pytorch/pull/148834

janeyx99 reviewed (commented):
looks good

desertfire committed:
Update on "[AOTI] Remove aoti_torch_cpu__weight_int4pack_mm_cpu_tensor"


Summary: shim.h is only meant for generic tensor util shim functions. Switch to use the auto fallback generation.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Remove aoti_torch_cpu__weight_int4pack_mm_cpu_tensor"


Summary: shim.h is only meant for generic tensor util shim functions. Switch to use the auto fallback generation.

[ghstack-poisoned]

janeyx99 commented:
ah nice you found the kernel for the fallback

desertfire commented:
The new version still doesn't work due to the way _weight_int4pack_mm_for_cpu is declared, https://www.internalfb.com/code/fbsource/[d5d3b559397b]/fbcode/caffe2/aten/src/ATen/native/native_functions.yaml?lines=4177. qGroupSize being an integer should be put at the end of the arg list!  @Xia-Weiwen , the current status is quite broken. We are tightening c/shim.h and trying to make sure c/shim.h is does not contain fallback ops. At this point, I want to just delete that function from c/shim.h, and disable the corresponding cpp_wrapper tests.  cc @EikanWang

Xia-Weiwen commented:
@desertfire Sorry but I am quite confused now. The declaration of `_weight_int4pack_mm_for_cpu` is not good for max-autotune since it has a scalar argument. That's why I added a wrapper op `_weight_int4pack_mm_cpu_tensor` whose arguments are all tensors. And to make it work with Inductor cpp wrapper, I made some changes in torch/csrc/inductor/aoti_torch/c/shim.h. And per your comment https://github.com/pytorch/pytorch/pull/145250/files#r1954568031, I thought I only needed to remove the changes in shim.h and to add the op to fallback_ops.py. But now we have this PR. May I know what I am supposed to do? Thanks.

desertfire committed:
Update on "[AOTI] Remove aoti_torch_cpu__weight_int4pack_mm_cpu_tensor"


Summary: shim.h is only meant for generic tensor util shim functions. We should switch to use the auto fallback generation, but it will need some extra care on the op schema.

[ghstack-poisoned]

Xia-Weiwen commented:
> 1. You can check out the current version to see why my fix doesn't work.
> 2. Max-autotune doesn't require an op to have all args as tensors, see convolution as an example, https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/conv.py . For schema change, you can refer to [[Inductor][Quant] Change the schema of QLinear Binary #129049](https://github.com/pytorch/pytorch/pull/129049) as an eample.
> 3. The change I am going to make doesn't affect the regular torch.compile flow. You can take time to work on fixing the cpp-wrapper side.

Thanks. I will try suggestion 1. For 2, my statement was inaccurate. It is not max-autotune but CPP GEMM template that requires tensor arguments. For 3, that's how I added the wrapper function. The original op is declared in native_functions.yaml, which is recommended to modify.

Xia-Weiwen commented:
Hi @desertfire I have checked out the latest main branch but I still don't know what issue you met and how to repro that. Could you please let me know? Thanks.

desertfire commented:
You can take my changes in https://github.com/pytorch/pytorch/commit/94268f83d4f506e69333d8e96d4e1f0b03ed8244, and then run `pytest test/inductor/test_cpu_cpp_wrapper.py -k test_woq_int4`.

desertfire commented:
@pytorchbot merge

Xia-Weiwen commented:
Hi @desertfire I can repro the issue now. But I don't know how to make it work with cpp wrapper without changing shim.h/.cpp. Do you have any suggestions? Thanks.

desertfire commented:
I can think of a few possible ways to solve it,
1. Fix CPP GEMM template to support scalar arg and reorder your schema arg order.
2. Update your schema to declare qGroupSize as tensor.
3. Move it to shim_mkldnn.cpp, and add codegen logic to insert extra include for "torch/csrc/inductor/aoti_torch/c/shim_mkldnn.h". 

1 and 2 are cleaner, but are also BC breaking. 3 is my least favorite, but I am fine with it since the shim function is isolated in shim_mkldnn.h.

desertfire commented:
@pytorchbot merge -f "Only affects CPU tests"

Xia-Weiwen commented:
Hi @desertfire After discussion with @EikanWang , we decided to fix it with two steps:
1. Move `aoti_torch_cpu__weight_int4pack_mm_cpu_tensor` to shim_mkldnn.cpp
2. Rename shim_mkldnn.cpp to shim_cpu.cpp (There are shim_cuda.cpp and shim_xpu.cpp. The name shim_mkldnn looks different.)

How does that sound to you? Thanks.

desertfire commented:
SGTM
